File: code/programs/typescript/unix-tools/src/touch.ts
/**
 * touch -- change file timestamps.
 *
 * === What This Program Does ===
 *
 * This is a reimplementation of the GNU `touch` utility in TypeScript. It
 * updates the access and modification times of each FILE to the current
 * time. If the file does not exist, it is created (unless -c is given).
 *
 * === How touch Works ===
 *
 * touch is most commonly used to create empty files:
 *
 *     touch newfile.txt         =>   creates newfile.txt (empty)
 *     touch existing.txt        =>   updates timestamps
 *     touch -c noexist.txt      =>   does nothing (file doesn't exist)
 *
 * === Time Selection ===
 *
 * By default, touch updates both access time (atime) and modification
 * time (mtime). You can select just one:
 *
 *     touch -a file.txt         =>   update only access time
 *     touch -m file.txt         =>   update only modification time
 *
 * === Custom Timestamps ===
 *
 * Instead of the current time, you can specify a time with:
 * - `-d STRING`: Parse a date string (e.g., "2024-01-15 10:30:00")
 * - `-r FILE`: Use another file's timestamps
 * - `-t STAMP`: Use [[CC]YY]MMDDhhmm[.ss] format
 *
 * These three options are mutually exclusive.
 *
 * @module touch
 */

import * as fs from "node:fs";
import * as path from "node:path";
import { fileURLToPath } from "node:url";

// ---------------------------------------------------------------------------
// Import CLI Builder.
// ---------------------------------------------------------------------------

import { Parser } from "@coding-adventures/cli-builder";

// ---------------------------------------------------------------------------
// Locate the JSON spec file.
// ---------------------------------------------------------------------------

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const SPEC_FILE = path.resolve(__dirname, "..", "touch.json");

// ---------------------------------------------------------------------------
// Business Logic: Parse timestamp formats.
// ---------------------------------------------------------------------------

/**
 * Parse a -t timestamp in [[CC]YY]MMDDhhmm[.ss] format.
 *
 * The format is positional:
 *   - CC: century (e.g., 20)
 *   - YY: year within century (e.g., 24)
 *   - MM: month (01-12)
 *   - DD: day (01-31)
 *   - hh: hour (00-23)
 *   - mm: minute (00-59)
 *   - .ss: seconds (00-59), optional
 *
 * If CC is omitted, the century is inferred from YY:
 *   - YY in 69-99 => 19YY
 *   - YY in 00-68 => 20YY
 *
 * If CCYY is omitted entirely, the current year is used.
 */
function parseTimestamp(stamp: string): Date {
  // Separate seconds if present.
  let mainPart = stamp;
  let seconds = 0;

  const dotIndex = stamp.indexOf(".");
  if (dotIndex !== -1) {
    seconds = parseInt(stamp.substring(dotIndex + 1), 10);
    mainPart = stamp.substring(0, dotIndex);
  }

  let year: number;
  let month: number;
  let day: number;
  let hour: number;
  let minute: number;

  if (mainPart.length === 8) {
    // MMDDhhmm -- no year, use current year.
    year = new Date().getFullYear();
    month = parseInt(mainPart.substring(0, 2), 10);
    day = parseInt(mainPart.substring(2, 4), 10);
    hour = parseInt(mainPart.substring(4, 6), 10);
    minute = parseInt(mainPart.substring(6, 8), 10);
  } else if (mainPart.length === 10) {
    // YYMMDDhhmm -- two-digit year.
    const yy = parseInt(mainPart.substring(0, 2), 10);
    year = yy >= 69 ? 1900 + yy : 2000 + yy;
    month = parseInt(mainPart.substring(2, 4), 10);
    day = parseInt(mainPart.substring(4, 6), 10);
    hour = parseInt(mainPart.substring(6, 8), 10);
    minute = parseInt(mainPart.substring(8, 10), 10);
  } else if (mainPart.length === 12) {
    // CCYYMMDDhhmm -- four-digit year.
    year = parseInt(mainPart.substring(0, 4), 10);
    month = parseInt(mainPart.substring(4, 6), 10);
    day = parseInt(mainPart.substring(6, 8), 10);
    hour = parseInt(mainPart.substring(8, 10), 10);
    minute = parseInt(mainPart.substring(10, 12), 10);
  } else {
    throw new Error(`invalid timestamp format: '${stamp}'`);
  }

  return new Date(year, month - 1, day, hour, minute, seconds);
}

// ---------------------------------------------------------------------------
// Main: parse args via CLI Builder, then touch files.
// ---------------------------------------------------------------------------

function main(): void {
  // --- Step 1: Parse arguments ---------------------------------------------

  let result;

  try {
    const parser = new Parser(SPEC_FILE, process.argv);
    result = parser.parse();
  } catch (err: unknown) {
    if (err && typeof err === "object" && "errors" in err) {
      const errors = (err as { errors: Array<{ message: string }> }).errors;
      for (const error of errors) {
        process.stderr.write(`touch: ${error.message}\n`);
      }
      process.exit(1);
    }
    throw err;
  }

  // --- Step 2: Dispatch on result type -------------------------------------

  if ("text" in result) {
    process.stdout.write(result.text + "\n");
    process.exit(0);
  }

  if ("version" in result && !("flags" in result)) {
    process.stdout.write(result.version + "\n");
    process.exit(0);
  }

  // --- Step 3: Extract flags and arguments ---------------------------------

  const flags = (result as { flags: Record<string, unknown> }).flags;
  const args = (result as { arguments: Record<string, unknown> }).arguments;

  const accessOnly = !!flags["access_only"];
  const modificationOnly = !!flags["modification_only"];
  const noCreate = !!flags["no_create"];
  const dateStr = flags["date"] as string | undefined;
  const referenceFile = flags["reference"] as string | undefined;
  const timestampStr = flags["timestamp"] as string | undefined;

  // Normalize file list.
  let files = args["files"] as string[] | string;
  if (typeof files === "string") {
    files = [files];
  }

  // --- Step 4: Determine the timestamp to use ------------------------------
  // By default, use the current time.
  // -d, -r, and -t are mutually exclusive alternatives.

  let targetTime = new Date();

  if (dateStr) {
    targetTime = new Date(dateStr);
    if (isNaN(targetTime.getTime())) {
      process.stderr.write(`touch: invalid date format '${dateStr}'\n`);
      process.exit(1);
    }
  } else if (referenceFile) {
    try {
      const stat = fs.statSync(referenceFile);
      targetTime = stat.mtime;
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : String(err);
      process.stderr.write(`touch: ${referenceFile}: ${message}\n`);
      process.exit(1);
    }
  } else if (timestampStr) {
    try {
      targetTime = parseTimestamp(timestampStr);
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : String(err);
      process.stderr.write(`touch: ${message}\n`);
      process.exit(1);
    }
  }

  // --- Step 5: Touch each file ---------------------------------------------

  for (const file of files) {
    const exists = fs.existsSync(file);

    if (!exists) {
      if (noCreate) {
        // -c: don't create, just skip.
        continue;
      }

      // Create the file.
      try {
        fs.writeFileSync(file, "");
      } catch (err: unknown) {
        const message = err instanceof Error ? err.message : String(err);
        process.stderr.write(`touch: ${file}: ${message}\n`);
        process.exitCode = 1;
        continue;
      }
    }

    // Update timestamps.
    try {
      const stat = fs.statSync(file);

      // Determine atime and mtime.
      // If neither -a nor -m is given, update both.
      // If -a is given, update only atime (keep mtime).
      // If -m is given, update only mtime (keep atime).
      const newAtime = accessOnly || (!accessOnly && !modificationOnly)
        ? targetTime
        : stat.atime;
      const newMtime = modificationOnly || (!accessOnly && !modificationOnly)
        ? targetTime
        : stat.mtime;

      fs.utimesSync(file, newAtime, newMtime);
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : String(err);
      process.stderr.write(`touch: ${file}: ${message}\n`);
      process.exitCode = 1;
    }
  }
}

// ---------------------------------------------------------------------------
// Run the program.
// ---------------------------------------------------------------------------

main();

File: code/programs/elixir/unix-tools/lib/unix_tools/touch.ex
defmodule UnixTools.Touch do
  @moduledoc """
  touch -- change file timestamps.

  ## What This Program Does

  This is a reimplementation of the GNU `touch` utility in Elixir. It updates
  the access and modification times of each FILE to the current time. If the
  file does not exist, it is created (unless -c is given).

  ## How touch Works

  touch is most commonly used to create empty files:

      touch newfile.txt         =>   creates newfile.txt (empty)
      touch existing.txt        =>   updates timestamps
      touch -c noexist.txt      =>   does nothing (file doesn't exist)

  ## Time Selection

  By default, touch updates both access time and modification time. You can
  select just one with `-a` (access only) or `-m` (modification only).

  ## Custom Timestamps

  Instead of the current time, you can specify a time with:
  - `-d STRING`: Parse a date string.
  - `-r FILE`: Use another file's timestamps.
  - `-t STAMP`: Use [[CC]YY]MMDDhhmm[.ss] format.
  """

  alias CodingAdventures.CliBuilder.Parser
  alias CodingAdventures.CliBuilder.{ParseResult, HelpResult, VersionResult, ParseErrors}

  # ---------------------------------------------------------------------------
  # Entry point
  # ---------------------------------------------------------------------------

  @doc """
  Entry point. Receives `argv` as a list of strings.
  """
  def main(argv) do
    spec_path = resolve_spec_path()

    case Parser.parse(spec_path, ["touch" | argv]) do
      {:ok, %HelpResult{text: text}} ->
        IO.puts(text)

      {:ok, %VersionResult{version: version}} ->
        IO.puts(version)

      {:ok, %ParseResult{flags: flags, arguments: arguments}} ->
        access_only = !!flags["access_only"]
        modification_only = !!flags["modification_only"]
        no_create = !!flags["no_create"]
        date_str = flags["date"]
        reference_file = flags["reference"]
        timestamp_str = flags["timestamp"]

        file_list = normalize_files(arguments["files"])

        # Determine the target time.
        target_time = determine_time(date_str, reference_file, timestamp_str)

        Enum.each(file_list, fn file ->
          touch_file(file, target_time, access_only, modification_only, no_create)
        end)

      {:error, %ParseErrors{errors: errors}} ->
        Enum.each(errors, fn e ->
          IO.puts(:stderr, "touch: #{e.message}")
        end)

        System.halt(1)
    end
  end

  # ---------------------------------------------------------------------------
  # Business Logic: Determine timestamp and touch files.
  # ---------------------------------------------------------------------------

  @doc """
  Determine the timestamp to use based on flags.

  Returns a DateTime or NaiveDateTime representing the desired time.
  If no time-source flag is given, uses the current time.
  """
  def determine_time(nil, nil, nil), do: NaiveDateTime.utc_now()

  def determine_time(date_str, nil, nil) when is_binary(date_str) do
    case NaiveDateTime.from_iso8601(date_str) do
      {:ok, dt} -> dt
      {:error, _} ->
        # Try parsing as a date only.
        case Date.from_iso8601(date_str) do
          {:ok, d} -> NaiveDateTime.new!(d, ~T[00:00:00])
          {:error, _} ->
            IO.puts(:stderr, "touch: invalid date format '#{date_str}'")
            System.halt(1)
        end
    end
  end

  def determine_time(nil, reference_file, nil) when is_binary(reference_file) do
    case File.stat(reference_file, time: :posix) do
      {:ok, stat} ->
        DateTime.from_unix!(stat.mtime) |> DateTime.to_naive()

      {:error, reason} ->
        IO.puts(:stderr, "touch: #{reference_file}: #{:file.format_error(reason)}")
        System.halt(1)
    end
  end

  def determine_time(nil, nil, timestamp_str) when is_binary(timestamp_str) do
    parse_timestamp(timestamp_str)
  end

  @doc """
  Parse a -t timestamp in [[CC]YY]MMDDhhmm[.ss] format.

  The format is positional and the century is inferred if omitted.
  """
  def parse_timestamp(stamp) do
    # Separate seconds if present.
    {main_part, seconds} =
      case String.split(stamp, ".") do
        [main, secs] -> {main, String.to_integer(secs)}
        [main] -> {main, 0}
      end

    {year, month, day, hour, minute} =
      case String.length(main_part) do
        8 ->
          # MMDDhhmm -- use current year.
          {DateTime.utc_now().year,
           String.slice(main_part, 0, 2) |> String.to_integer(),
           String.slice(main_part, 2, 2) |> String.to_integer(),
           String.slice(main_part, 4, 2) |> String.to_integer(),
           String.slice(main_part, 6, 2) |> String.to_integer()}

        10 ->
          # YYMMDDhhmm -- two-digit year.
          yy = String.slice(main_part, 0, 2) |> String.to_integer()
          full_year = if yy >= 69, do: 1900 + yy, else: 2000 + yy

          {full_year,
           String.slice(main_part, 2, 2) |> String.to_integer(),
           String.slice(main_part, 4, 2) |> String.to_integer(),
           String.slice(main_part, 6, 2) |> String.to_integer(),
           String.slice(main_part, 8, 2) |> String.to_integer()}

        12 ->
          # CCYYMMDDhhmm -- four-digit year.
          {String.slice(main_part, 0, 4) |> String.to_integer(),
           String.slice(main_part, 4, 2) |> String.to_integer(),
           String.slice(main_part, 6, 2) |> String.to_integer(),
           String.slice(main_part, 8, 2) |> String.to_integer(),
           String.slice(main_part, 10, 2) |> String.to_integer()}

        _ ->
          IO.puts(:stderr, "touch: invalid timestamp format '#{stamp}'")
          System.halt(1)
      end

    NaiveDateTime.new!(year, month, day, hour, minute, seconds)
  end

  @doc """
  Touch a single file: create if needed, then update timestamps.
  """
  def touch_file(file, target_time, access_only, modification_only, no_create) do
    file_exists = File.exists?(file)

    if not file_exists do
      if no_create do
        # -c: don't create, just skip.
        :ok
      else
        case File.write(file, "") do
          :ok -> update_timestamps(file, target_time, access_only, modification_only)
          {:error, reason} ->
            IO.puts(:stderr, "touch: #{file}: #{:file.format_error(reason)}")
        end
      end
    else
      update_timestamps(file, target_time, access_only, modification_only)
    end
  end

  defp update_timestamps(file, target_time, access_only, modification_only) do
    # Convert NaiveDateTime to posix seconds for File.touch.
    epoch = NaiveDateTime.diff(target_time, ~N[1970-01-01 00:00:00])

    # Determine which times to update.
    update_atime = access_only or (not access_only and not modification_only)
    update_mtime = modification_only or (not access_only and not modification_only)

    case File.stat(file, time: :posix) do
      {:ok, stat} ->
        new_atime = if update_atime, do: epoch, else: stat.atime
        new_mtime = if update_mtime, do: epoch, else: stat.mtime

        # Use :file.change_time to set both timestamps.
        :file.change_time(to_charlist(file), new_atime, new_mtime)

      {:error, reason} ->
        IO.puts(:stderr, "touch: #{file}: #{:file.format_error(reason)}")
    end
  end

  # ---------------------------------------------------------------------------
  # Helpers
  # ---------------------------------------------------------------------------

  defp normalize_files(files) when is_list(files), do: files
  defp normalize_files(file) when is_binary(file), do: [file]

  defp resolve_spec_path do
    candidates = [
      if(function_exported?(Mix, :Project, 0),
        do: Path.join(Mix.Project.config()[:lockfile] |> Path.dirname(), "touch.json"),
        else: nil
      ),
      "touch.json",
      Path.join(:code.priv_dir(:unix_tools) |> to_string(), "touch.json")
    ]

    Enum.find(candidates, fn
      nil -> false
      candidate_path -> File.exists?(candidate_path)
    end) ||
      raise "Could not find touch.json spec file"
  end
end
